substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java

```diff
@@ -0,0 +1,756 @@
+                    DwarfRegEncodingAArch64.ZR.encoding,
+                    DwarfRegEncodingAArch64.SP.encoding,
+                    DwarfRegEncodingAArch64.V0.encoding,
+                    DwarfRegEncodingAArch64.V1.encoding,
+                    DwarfRegEncodingAArch64.V2.encoding,
+                    DwarfRegEncodingAArch64.V3.encoding,
+                    DwarfRegEncodingAArch64.V4.encoding,
+                    DwarfRegEncodingAArch64.V5.encoding,
+                    DwarfRegEncodingAArch64.V6.encoding,
+                    DwarfRegEncodingAArch64.V7.encoding,
+                    DwarfRegEncodingAArch64.V8.encoding,
+                    DwarfRegEncodingAArch64.V9.encoding,
+                    DwarfRegEncodingAArch64.V10.encoding,
+                    DwarfRegEncodingAArch64.V11.encoding,
+                    DwarfRegEncodingAArch64.V12.encoding,
+                    DwarfRegEncodingAArch64.V13.encoding,
+                    DwarfRegEncodingAArch64.V14.encoding,
+                    DwarfRegEncodingAArch64.V15.encoding,
+                    DwarfRegEncodingAArch64.V16.encoding,
+                    DwarfRegEncodingAArch64.V17.encoding,
+                    DwarfRegEncodingAArch64.V18.encoding,
+                    DwarfRegEncodingAArch64.V19.encoding,
+                    DwarfRegEncodingAArch64.V20.encoding,
+                    DwarfRegEncodingAArch64.V21.encoding,
+                    DwarfRegEncodingAArch64.V22.encoding,
+                    DwarfRegEncodingAArch64.V23.encoding,
+                    DwarfRegEncodingAArch64.V24.encoding,
+                    DwarfRegEncodingAArch64.V25.encoding,
+                    DwarfRegEncodingAArch64.V26.encoding,
+                    DwarfRegEncodingAArch64.V27.encoding,
+                    DwarfRegEncodingAArch64.V28.encoding,
+                    DwarfRegEncodingAArch64.V29.encoding,
+                    DwarfRegEncodingAArch64.V30.encoding,
+                    DwarfRegEncodingAArch64.V31.encoding,
+    };
+
+    // register numbers used by DWARF for AMD64 registers
+    public enum DwarfRegEncodingAMD64 {
+        RAX(0),
+        RDX(1),
+        RCX(2),
+        RBX(3),
+        RSI(4),
+        RDI(5),
+        RBP(6),
+        RSP(7),
+        R8(8),
+        R9(9),
+        R10(10),
+        R11(11),
+        R12(12),
+        R13(13),
+        R14(14),
+        R15(15),
+        XMM0(17),
+        XMM1(18),
+        XMM2(19),
+        XMM3(20),
+        XMM4(21),
+        XMM5(22),
+        XMM6(23),
+        XMM7(24),
+        XMM8(25),
+        XMM9(26),
+        XMM10(27),
+        XMM11(28),
+        XMM12(29),
+        XMM13(30),
+        XMM14(31),
+        XMM15(32);
+
+        public final int encoding;
+
+        DwarfRegEncodingAMD64(int encoding) {
+            this.encoding = encoding;
+        }
+    }
+
+    // map from compiler X86_64 register indices to corresponding dwarf AMD64 register index
+    private static final int[] GRAAL_X86_64_TO_DWARF_REG_MAP = {
```

Comment:
Use the same pattern here as shown in https://github.com/oracle/graal/pull/4505/files#r865995904

Reply:
This doesn't actually work -- it does for AArch64 but only because that architecture has a more sensible naming scheme.

The AArch64 and X86_64 enum tags are sorted by the DWARF encoding. So, for x86_64 the register sequence for DWARF 0...7 is RAX(0),  RDX(1),  RCX(2), RBX(3), RSI(4), RDI(5), RBP(6), RSP(7).

The Graal register numbers do not follow the same order for their register  numbering. For numberings 0...7 the corresponding registers are RAX(0), RCX(1), RDX(2), RBX(3), RSP(4), RBP(5), RSI(6), RDI(7).

That different order is taken account of in the sequence in which the DwarfRegEncoding tags appear in the array initialization expression (notice how the first 8 entries in the following table are permuted). What it boils down to is that there are actually two translation steps at play in this iniitalization. The Graal register number n (range 0-31) indexes an array slot populated using a field expression that accesses the corresponding DwarfRegEncoding enum tag. The field retrieves the DWARF encoding for that DwarfRegEncoding.

This problem could be finessed by re-ordering the enum tags so that they are declared out of DWARF encoding order. If that were done then the iterative enumeration you are suggesting could be made to work. However, I think that would make this already fairly obscure translation process more difficult to follow.

I'd prefer to keep the current layout and add a comment to explain the rationale for this two specific step translation. As it stands, the code explicitly and independently displays each of the two steps in the translation via i) the argument passed to the enum tag and ii) the order in which entries appear in the initializer. Relying on the tag ordering to determine the permutation makes this step implicit and my view is that it would make the details of the mapping harder to understand and to document.

Reply:
I have pushed a patch that retains the current array initilaizer and documents the steps in the translation. I hope that is ok.

Reply:
> Relying on the tag ordering to determine the permutation makes this step implicit and my view is that it would make the details of the mapping harder to understand and to document.

Alternatively you could just add another int field to the enum that holds the corresponding Graal register number explicitly for a given enum value (e.g. `regNumGraal`). Then again turning this into `GRAAL_X86_64_TO_DWARF_REG_MAP` is just a single line of code (similar to what I suggested before).

Reply:
Yes, that would be much better. I'll push another update.

Reply:
Ok, I did as you suggested.

n.b. To make this more bullet-proof I used the named Register instances {r0, ... v31} and {rax ... xmm15} from, respectively, AArch64 and AMD64 to specify the register numbers passed to the corresponding DwarfRegister enum tag. That required adding a dependency to the python suites exposing package jdk.interal.vm.ci/jdk.vm.ci.code to module com.oracle.objectfile . I hope that is ok.

Reply:
Repushed with format fixes.